clang/lib/Parse/ParseContracts.cpp

```diff
@@ -0,0 +1,437 @@
+//
+///   function-contract-specifier:
+///       precondition-specifier
+///       postcondition-specifier
+//
+///   precondition-specifier:
+///       pre attribute-specifier-seq[opt] ( conditional-expression )
+///
+///   postcondition-specifier:
+///       post attribute-specifier-seq[opt] ( result-name-introducer[opt]
+///       conditional-expression )
+///
+///   result-name-introducer:
+///       attributed-identifier :
+void Parser::ParseContractSpecifierSequence(Declarator &DeclarationInfo,
+                                            bool EnterScope,
+                                            QualType TrailingReturnType) {
+  if (!isFunctionContractKeyword(Tok))
+    return;
+
+  std::optional<QualType> CachedType;
+  auto ReturnTypeResolver = [&]() {
+    if (!CachedType) {
+      QualType ReturnType = TrailingReturnType;
+      if (ReturnType.isNull()) {
+        TypeSourceInfo *TInfo = Actions.GetTypeForDeclarator(DeclarationInfo);
+        assert(TInfo && TInfo->getType()->isFunctionType());
+        ReturnType = TInfo->getType()->getAs<FunctionType>()->getReturnType();
+      }
+      CachedType = ReturnType;
+    }
+    return CachedType.value();
+  };
+  std::optional<ParseScope> ParserScope;
+
+  std::optional<Sema::CXXThisScopeRAII> ThisScope;
+  std::optional<Sema::FunctionScopeRAII> PopFnContext;
+
+  if (EnterScope) {
+    ParserScope.emplace(this, Scope::DeclScope | Scope::FunctionPrototypeScope |
+                                  Scope::FunctionDeclarationScope);
+
+   // PopFnContext.emplace(Actions);
+   // Actions.PushFunctionScope();
+
+    auto FTI = DeclarationInfo.getFunctionTypeInfo();
+
+    for (unsigned i = 0; i != FTI.NumParams; ++i) {
+      ParmVarDecl *Param = cast<ParmVarDecl>(FTI.Params[i].Param);
+      Actions.ActOnReenterCXXMethodParameter(getCurScope(), Param);
+    }
+  }
+
+  InitCXXThisScopeForDeclaratorIfRelevant(DeclarationInfo,
+                                          DeclarationInfo.getDeclSpec(),
+                                          ThisScope);
+  bool IsInvalid = false;
+  SourceLocation StartLoc = Tok.getLocation();
+
+  SmallVector<ContractStmt *, 4> Contracts;
+  while (isFunctionContractKeyword(Tok)) {
+    bool IsInvalidTmp = false;
+    StmtResult Contract =
+        ParseFunctionContractSpecifierImpl(ReturnTypeResolver, EnterScope ? CSO_ParentContext : CSO_FunctionContext, IsInvalidTmp);
+    IsInvalid |= IsInvalidTmp;
+    if (Contract.isUsable())
+      Contracts.push_back(Contract.getAs<ContractStmt>());
+  }
+  ContractSpecifierDecl *Seq = Actions.ActOnFinishContractSpecifierSequence(
+      Contracts, StartLoc, IsInvalid);
+
+  assert(DeclarationInfo.Contracts == nullptr && "Already have contracts?");
+
+  DeclarationInfo.Contracts = Seq;
+}
+
+StmtResult Parser::ParseFunctionContractSpecifierImpl(
+    llvm::function_ref<QualType()> ReturnTypeResolver, ContractScopeOffset ScopeOffset, bool &IsInvalid) {
```

Comment:
Can we 
 -  call that `ParseContractAfterKeyword` or something
 -  Pass in parameter the kind (pre/post/assert) and the kw location?

it would let us not mix the handling of the assert kw and that of pre/post - they appear in different contexts and work differently